slash/shared/src/main/scala/ai/dragonfly/math/vector/package.scala

```diff
@@ -198,54 +204,59 @@ package object vector {
         thisVector.zip(thatVector).map{ case (thisV, thatV) => (thisV - μThis) * (thatV - μThat) }.sum / (thisVector.size -1)
       }
 
-      def pearsonCorrelationCoefficient(thatVector: Vec[N]): Double = {
+      inline def pearsonCorrelationCoefficient(thatVector: Vec[N]): Double = {
         val n = thisVector.size
-        val sum_x = thisVector.sum
-        val sum_y = thatVector.sum
-        val sum_xy = thisVector.zip(thatVector).map{ case (thisV, thatV) => thisV * thatV }.sum
-        val sum_x2 = thisVector.map(squareInPlace(_)).sum
-        val sum_y2 = thatVector.map(squareInPlace(_)).sum
+        var i = 0
+
+        var sum_x = 0.0
+        var sum_y = 0.0
+        var sum_xy = 0.0
+        var sum_x2 = 0.0
+        var sum_y2 = 0.0
+
+        while (i < n) {
+          sum_x = sum_x + thisVector(i)
+          sum_y = sum_y + thatVector(i)
+          sum_xy = sum_xy + thisVector(i) * thatVector(i)
+          sum_x2 = sum_x2 + squareInPlace(thisVector(i))
+          sum_y2 = sum_y2 + squareInPlace(thatVector(i))
+          i = i + 1
+        }
         (n * sum_xy - (sum_x * sum_y)) / Math.sqrt( (sum_x2 * n - squareInPlace(sum_x)) * (sum_y2 * n - squareInPlace(sum_y)) )
       }
 
-      def spearmansRankCorrelation(thatVector: Vec[N]): Double = {
+      inline def spearmansRankCorrelation(thatVector: Vec[N]) : Double =
         val theseRanks = thisVector.elementRanks
         val thoseRanks = thatVector.elementRanks
-        val diffs = theseRanks - thoseRanks
-        val diffs_2 = diffs.map(squareInPlace(_))
-        val n = theseRanks.size
-        val s = diffs_2.sum
-        val numerator = 6 * diffs_2.sum
-        val denominator = n * (squareInPlace(n) - 1)
-        1 - ( numerator / denominator)
-      }
+        theseRanks.pearsonCorrelationCoefficient(thoseRanks)
+
 
       // An alias - pearson is the most commonly requested type of correlation
-      def corr(thatVector: Vec[N]): Double = pearsonCorrelationCoefficient(thatVector)
+      inline def corr(thatVector: Vec[N]): Double = pearsonCorrelationCoefficient(thatVector)
 
-      def elementRanks: Vec[N] = {
+      inline def elementRanks: Vec[N] = {
         val (sorted, originalPosition) = thisVector.zipWithIndex.toVector.sortBy(_._1).unzip
         val ranks : Vec[N] = NArray.tabulate[Double](thisVector.dimension)(i => (i+1).toDouble)
 
         var currentValue = sorted(0)
         var i = 0
         var currentSum = 0.0
         var currentCount = 0
-        var resultList = List[Double]()
+        var resultList = ListBuffer[Double]()
```

Comment:
Now that `resultList` is a mutable `ListBuffer` that gets modified in-place, it could be declared as `val`.

Reply:
Could we remove `List` and `ListBuffer` altogether?

Reply:
I'll sneak the `var` -> `val` change into a future PR. 

It would also be possible to remove the List, ListBuffer construct altogether, I agree. I'm not sure how much the perf benefit would be though, and it does add complexity as I think it would hide the intent a little.